packages/node-opentelemetry/src/context.ts

```diff
@@ -0,0 +1,114 @@
+import { Attributes, context, trace } from '@opentelemetry/api';
+import { Span } from '@opentelemetry/sdk-trace-base';
+import { suppressTracing } from '@opentelemetry/core';
+
+import hdx from './debug';
+
+const HDX_CONTEXT_MAX_SPANS_PER_TRACE = 100;
+const HDX_CONTEXT_MAX_TRACKED_TRACE_IDS = 50000; // ~ 500 MB of memory (10 spans / trace, 1k per span)
+const HDX_CONTEXT_REFRESHER_INTERVAL = 10000;
+const HDX_CONTEXT_TRACE_ATTRIBUTES_EXPIRATION = 5 * 60 * 1000;
+
+class HyperDXContext {
+  private readonly _traceAttributes = new Map<string, Attributes>();
+  private readonly _traceMap = new Map<
+    string,
+    {
+      spans: Span[];
+      lastUpdateAt: number;
+    }
+  >();
+
+  constructor() {
+    // expires after 5 minutes
+    setInterval(() => {
+      hdx(`Running _traceMap expiration check`);
+      const now = Date.now();
+      for (const [traceId, data] of this._traceMap.entries()) {
+        if (now - data.lastUpdateAt > HDX_CONTEXT_TRACE_ATTRIBUTES_EXPIRATION) {
+          this._traceMap.delete(traceId);
+          this._traceAttributes.delete(traceId);
+          hdx(`Deleted traceId ${traceId} from _traceMap and _traceAttributes`);
+        }
+      }
+    }, HDX_CONTEXT_REFRESHER_INTERVAL);
+  }
+
+  private _getActiveSpanTraceId = (): string | undefined => {
+    const activeSpan = trace.getActiveSpan();
+    if (!activeSpan) {
+      return undefined;
+    }
+    return activeSpan.spanContext().traceId;
+  };
+
+  private _setTraceAttributesForAllSpans = (traceId: string): void => {
+    const attributes = this.getTraceAttributes(traceId);
+    if (!attributes) {
+      return;
+    }
+    const traceData = this._traceMap.get(traceId);
+    if (traceData) {
+      for (const span of traceData.spans) {
+        if (!span.ended) {
+          span.setAttributes(attributes);
+        }
+      }
+    }
+  };
+
+  addTraceSpan = (traceId: string, span: Span): void => {
+    if (this._traceMap.size >= HDX_CONTEXT_MAX_TRACKED_TRACE_IDS) {
```

Comment:
Shouldn't we do this check when we create a new entry in the trace map? Otherwise I believe this will reject spans on existing trace ids if you're at the trace limit (though understood this is a pretty narrow edge case, and perhaps intentional?).

Reply:
Ah what did I do 🤦 ....let me fix this real quick